Source/Core/Absy.cs

```diff
@@ -2866,11 +2868,17 @@ public Function(IToken tok, string name, List<TypeVariable> typeParams, List<Var
         def = new ForallExpr(tok, quantifiedTypeVars, dummies,
                              kv,
                              new Trigger(tok, true, new List<Expr> { call }, null),
-                             def);
+                             def, false, isFuncDef);
       }
       DefinitionAxiom = new Axiom(tok, def);
       return DefinitionAxiom;
     }
+
+    public Axiom CreateFunctionDefinition(Expr definition, QKeyValue kv = null) {
+      CreateDefinitionAxiom(definition, kv, true);
+      return DefinitionAxiom;
+    }
+
```

Comment:
It seems we are creating an axiom even when the programmer is indicating a function definition using :define.  Is there some peculiarity of Boogie that is forcing this design?

Reply:
Yes.  At first, I tried an implementation that just stores the function definition as a special field in the Function class.  However, such functions must ultimately be translated to VCExprs. 
 This requires adding a new type of VCExpr for function definitions.  While conceptually clean, this ends up being a very heavy and error-prone change, as it requires changing all of the VCExpr visitors (there are like 10-20 of them).  I thought it was maybe not worth it when I could instead recover the function from the axiom quite easily and not have to touch the VCExpr class heirarchy.  I'm open to ideas if you think there is a better way.